packages/@webex/plugin-meetings/src/reachability/clusterReachability.ts

```diff
@@ -0,0 +1,325 @@
+      );
+
+      return undefined;
+    }
+  }
+
+  getResult() {
+    return this.result;
+  }
+
+  isUnreachable(protocol: 'udp' | 'tcp') {
+    if (protocol === 'udp') {
+      return this.numUdpUrls > 0 && this.result.udp.reachable !== 'true';
+    }
+
+    return this.numTcpUrls > 0 && this.result.tcp.reachable !== 'true';
+  }
+
+  private closePeerConnection() {
+    if (this.pc) {
+      this.pc.onicecandidate = null;
+      this.pc.onicegatheringstatechange = null;
+      this.pc.close();
+    }
+  }
+
+  private finishReachabilityCheck() {
+    this.defer.resolve();
+  }
+
+  /**
+   * Adds public IP (client media IPs)
+   * @param {string} protocol
+   * @param {string} publicIP
+   * @returns {void}
+   */
+  protected addPublicIP(protocol: 'udp' | 'tcp', publicIP?: string | null) {
+    const {CLOSED} = CONNECTION_STATE;
+
+    if (this.pc?.connectionState === CLOSED) {
+      LoggerProxy.logger.log(
+        `Reachability:index#addPublicIP --> Attempting to set publicIP of ${publicIP} on closed peerConnection.`
+      );
+    }
+
+    const result = this.result[protocol];
+
+    if (publicIP) {
+      if (result.clientMediaIPs) {
+        if (!result.clientMediaIPs.includes(publicIP)) {
+          result.clientMediaIPs.push(publicIP);
+        }
+      } else {
+        result.clientMediaIPs = [publicIP];
+      }
+    }
+  }
+
+  private registerIceGatheringStateChangeListener() {
+    this.pc.onicegatheringstatechange = () => {
+      const {COMPLETE} = ICE_GATHERING_STATE;
+
+      if (this.pc.iceConnectionState === COMPLETE) {
+        console.log(
+          `marcin: ${this.name}: this.pc.iceConnectionState=${this.pc.iceConnectionState}`
+        );
+        this.closePeerConnection();
+        this.finishReachabilityCheck();
+      }
+    };
+  }
+
+  private haveWeGotAllResults(): boolean {
+    const expecting = {
+      udp: this.numUdpUrls > 0,
+      tcp: this.numTcpUrls > 0,
+    };
+
+    const gotResult = {
+      udp: this.result.udp.reachable === 'true',
+      tcp: this.result.tcp.reachable === 'true',
+    };
+
+    return ['udp', 'tcp'].every((protocol) => expecting[protocol] === gotResult[protocol]);
+  }
+
+  private registerIceCandidateListener() {
+    this.pc.onicecandidate = (e) => {
+      const CANDIDATE_TYPES = {
+        SERVER_REFLEXIVE: 'srflx',
+        RELAY: 'relay',
+      };
+
+      if (e.candidate) {
+        if (e.candidate.type !== 'host') {
+          console.log('marcin: e.candidate', e.candidate);
+        }
+        if (e.candidate.type === CANDIDATE_TYPES.SERVER_REFLEXIVE) {
+          if (this.result.udp.latencyInMilliseconds === undefined) {
+            const elapsed = this.getElapsedTime();
+
+            LoggerProxy.logger.log(
+              // @ts-ignore
+              `Reachability:index#onIceCandidate --> Successfully reached ${this.name} over UDP: ${elapsed}ms`
+            );
+            this.result.udp.latencyInMilliseconds = elapsed.toString();
```

Comment:
needs to be a string?

Reply:
also does reflexive map to UDP, and relay to TCP?

Reply:
yes, it's weird but that's what the backend expects, there are comments about it in the type definition - see `TransportResult`, so I just keep the data always in that format. We could have our own type for all the internal logic and storing of the results and only convert to the type expected by the backend just before sending it. I don't mind it that much which way we do it. Do you think it would be better to change it?

Reply:
No, i missed the comment in the type -- if that's what the backend wants it's ok

Reply:
in the end I've decided to change it and have 2 separate types: 
`ReachabilityResults` - that's our internal type
`ReachabilityResultsForBackend` - that's the weird type that backend wants from us
gives us freedom to store the data however we want it